File: src/main/java/com/leijendary/spring/microservicetemplate/error/HttpMessageNotReadableExceptionHandler.java
package com.leijendary.spring.microservicetemplate.error;

import com.fasterxml.jackson.databind.JsonMappingException;
import com.fasterxml.jackson.databind.JsonMappingException.Reference;
import com.fasterxml.jackson.databind.exc.InvalidFormatException;
import com.leijendary.spring.microservicetemplate.data.response.ErrorResponse;
import lombok.RequiredArgsConstructor;
import org.springframework.context.MessageSource;
import org.springframework.core.annotation.Order;
import org.springframework.http.converter.HttpMessageNotReadableException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestControllerAdvice;

import java.util.List;

import static java.util.Locale.getDefault;
import static java.util.Optional.ofNullable;
import static java.util.stream.Collectors.joining;
import static org.springframework.http.HttpStatus.BAD_REQUEST;

@RestControllerAdvice
@Order(2)
@RequiredArgsConstructor
public class HttpMessageNotReadableExceptionHandler {

    private final MessageSource messageSource;

    @ExceptionHandler(HttpMessageNotReadableException.class)
    @ResponseStatus(BAD_REQUEST)
    public ErrorResponse catchHttpMessageNotReadable(final HttpMessageNotReadableException exception) {
        final var response = ErrorResponse.builder()
                .status(BAD_REQUEST);

        errors(exception, response);

        return response.build();
    }

    private void errors(final HttpMessageNotReadableException exception,
                        final ErrorResponse.ErrorResponseBuilder errorResponse) {
        final var source = "Request Body";
        final var code = "error.bad-request";
        var message = ofNullable(exception.getMessage())
                .orElse("");

        if (message.startsWith("Required request body is missing")) {
            errorResponse.addError(source, code, message.split(":")[0]);

            return;
        }

        if (exception.getCause() instanceof InvalidFormatException) {
            errors((InvalidFormatException) exception.getCause(), errorResponse);

            return;
        }

        if (exception.getCause() instanceof JsonMappingException) {
            errors((JsonMappingException) exception.getCause(), errorResponse);

            return;
        }

        message = ofNullable(exception.getMessage())
                .map(s -> s.replace("JSON decoding error: ", ""))
                .orElse(exception.getMessage());

        errorResponse.addError(source, code, message);
    }

    private void errors(final InvalidFormatException exception,
                        final ErrorResponse.ErrorResponseBuilder errorResponse) {
        final var path = getPath(exception.getPath());
        final var code = "error.invalid-format";
        final var message =  messageSource.getMessage(code,
                new Object[] { path, exception.getValue(), exception.getTargetType().getSimpleName() }, getDefault());

        errorResponse.addError(path, code, message);
    }

    private void errors(final JsonMappingException exception,
                            final ErrorResponse.ErrorResponseBuilder errorResponse) {
        errorResponse.addError(getPath(exception.getPath()), "error.invalid-format",
                exception.getOriginalMessage());
    }

    private String getPath(List<Reference> path) {
        return path.stream()
                .map(Reference::getFieldName)
                .collect(joining("."));
    }
}

File: src/main/kotlin/com/leijendary/spring/template/core/error/HttpMessageNotReadableExceptionHandler.kt
package com.leijendary.spring.template.core.error

import com.fasterxml.jackson.databind.JsonMappingException
import com.fasterxml.jackson.databind.JsonMappingException.Reference
import com.fasterxml.jackson.databind.exc.InvalidFormatException
import com.leijendary.spring.template.core.model.ErrorModel
import com.leijendary.spring.template.core.util.RequestContext.locale
import org.springframework.context.MessageSource
import org.springframework.core.annotation.Order
import org.springframework.http.HttpStatus.BAD_REQUEST
import org.springframework.http.converter.HttpMessageNotReadableException
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestControllerAdvice

private const val PARENT = "body"

@RestControllerAdvice
@Order(2)
class HttpMessageNotReadableExceptionHandler(private val messageSource: MessageSource) {
    @ExceptionHandler(HttpMessageNotReadableException::class)
    @ResponseStatus(BAD_REQUEST)
    fun catchHttpMessageNotReadable(exception: HttpMessageNotReadableException) = errors(exception)

    private fun errors(exception: HttpMessageNotReadableException): List<ErrorModel> {
        val source = listOf(PARENT)
        val code = "error.badRequest"
        var message = exception.message ?: ""
        val error: ErrorModel

        if (message.startsWith("Required request body is missing")) {
            error = ErrorModel(source, code, message.split(":".toRegex()).toTypedArray()[0])

            return listOf(error)
        }

        when (val cause = exception.cause) {
            is InvalidFormatException -> return errors(cause)
            is JsonMappingException -> return errors(cause)
        }

        message = exception.message?.replace("JSON decoding error: ", "") ?: ""
        error = ErrorModel(source, code, message)

        return listOf(error)
    }

    private fun errors(exception: InvalidFormatException): List<ErrorModel> {
        val sources = sources(exception.path)
        val code = "error.body.format.invalid"

        return sources.map {
            val field = it
                // Remove the parent field source
                .subList(1, it.size)
                .joinToString(".")
            val arguments = arrayOf(field, exception.value, exception.targetType.simpleName)
            val message = messageSource.getMessage(code, arguments, locale)

            ErrorModel(it, code, message)
        }
    }

    private fun errors(exception: JsonMappingException): List<ErrorModel> {
        val sources = sources(exception.path)
        val code = "error.body.format.invalid"
        val message = exception.originalMessage

        return sources.map { ErrorModel(it, code, message) }
    }

    private fun sources(path: List<Reference>): List<List<Any>> {
        return path.map {
            val source = mutableListOf<Any>(PARENT, it.fieldName)

            if (it.index >= 0) {
                source.add(1, it.index)
            }

            source
        }
    }
}
